src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.cs

```diff
@@ -165,5 +166,169 @@ public sealed override byte[] GetSubjectKeyIdentifier(X509Certificate2 certifica
                 return ski;
             }
         }
+
+        public override T GetPrivateKeyForSigning<T>(X509Certificate2 certificate, bool silent)
+        {
+            return GetPrivateKey<T>(certificate, silent, preferNCrypt: true);
+        }
+
+        public override T GetPrivateKeyForDecryption<T>(X509Certificate2 certificate, bool silent)
+        {
+            return GetPrivateKey<T>(certificate, silent, preferNCrypt: false);
+        }
+
+        private T GetPrivateKey<T>(X509Certificate2 certificate, bool silent, bool preferNCrypt) where T : AsymmetricAlgorithm
+        {
+            if (!certificate.HasPrivateKey)
+            {
+                return null;
+            }
+
+            SafeProvOrNCryptKeyHandle handle = GetCertificatePrivateKey(
+                certificate,
+                silent,
+                preferNCrypt,
+                out CryptKeySpec keySpec,
+                out Exception exception);
+
+            using (handle)
+            {
+                if (handle == null || handle.IsInvalid)
+                {
+                    if (exception != null)
+                    {
+                        throw exception;
+                    }
+
+                    return null;
+                }
+
+                if (keySpec == CryptKeySpec.CERT_NCRYPT_KEY_SPEC)
+                {
+                    using (SafeNCryptKeyHandle keyHandle = new SafeNCryptKeyHandle(handle.DangerousGetHandle(), handle))
+                    using (CngKey cngKey = CngKey.Open(keyHandle, CngKeyHandleOpenOptions.None))
+                    {
+                        if (typeof(T) == typeof(RSA))
+                            return (T)(object)new RSACng(cngKey);
+                        if (typeof(T) == typeof(ECDsa))
+                            return (T)(object)new ECDsaCng(cngKey);
+                        if (typeof(T) == typeof(DSA))
+                            return (T)(object)new DSACng(cngKey);
```

Comment:
I wonder if this (and similar sections) should be in some kind of Common file to make future supported algorithms easier? I feel like we've got lots of blocks like this scattered about.

Reply:
Well, this one is sort of unique, since it's playing with CNG pointers, and not many things should be doing that.

I agree about the meta point.  #25948 to track.